### ISSUE TITLE
End-of-session success display

### ISSUE THREAD

**Issue by KBMD** [jonkb/TicTrainer-node#29]
**This request should be near the top of the queue**--it was requested by OTs wanting to use this with their clients. 

At the end of each user session, display level, points and coins (and keep that on the screen).

At the end of each trainer session, display the user's level, points & coins, AND the info at the bottom of the session log, e.g.:
user l,p,c | 1,266,0
session length	220.507
levels gained	0
points earned	261
number of tics	4
longest tic free interval	128.945
number of 10s tic free intervals	19
report generated with TicTrainer version	2020.11.11

**Comment by jonkb** [comment]
This is talking about the normal TicTrainer interface, not the NT interface right?
Though maybe it'd be good for both?

**Comment by KBMD** [comment]
Yes, this is for the regular Tt RE-ERP sessions not the NT sessions.

Note:
For NT sessions, the nice feature would be on the "download log files" page
to have an option to download all log files for one day's work, which for
me it's essentially always identical to one admin-user pair, e.g. a1u55_* .
I.e. after a NT session I generally go to the "download log files" page,
select a given user e.g. u55, and then download each of the usually 6 files
that are now available via the left button in the 2nd row. It'd be nice to
do that in 1 click not 6+ clicks.

I'll add that 2nd paragraph as a feature request on GitHub.

Thank you! Remember to track your hours.
Love,
Dad

Kevin J. Black

**Comment by KBMD** [comment]
## update of this issue

### This request should be near the top of the queue--it was requested by OTs wanting to use this with their clients.

At the end of each **user** session, display level, points and coins (and keep that on the screen) along with personal best time i.e. longest total Tic-free time in mm:ss.

At the end of each **trainer** session, display the user's level, points & coins, AND the info at the bottom of the session log, e.g.:
user l,p,c | 1,266,0
personal best time 03:14   [i.e. longest total Tic-free time in mm:ss]
session length 220.507
levels gained 0
points earned 261
number of tics 4
longest tic free interval 128.945
number of 10s tic free intervals 19
report generated with TicTrainer version 2020.11.11

**Comment by jonkb** [comment]
I set this up in the "EOS_display" branch.
One thing I don't have yet is the user's personal best, since I haven't been keeping track of that number anywhere. It would be possible to extract by searching through every session file involving that user, but it'd probably make more sense to store a personal best time in the user's account.

**Comment by jonkb** [comment]
I also keep thinking of how I'd do things differently if I were building this from scratch...

- For example, I'm thinking of redoing the database as a MySQL database.
  - The task of finding "the set of users who have added this trainer to their list of trainers" ( #31 ) involves searching through every user's file for their list of linked trainers. If we were using a SQL database instead of individual files, then I think it'd be a single query. `select * from user_links where trainer_id in (L1, L2, L3, L4, L5);`
  - Also, at some point I really need to fix the fact that it's storing passwords in plain text.
- I've also been thinking of switching over to the express.js module, which makes lots of web server things simpler. It'd be really easy to set up persistant login sessions with express, for example. It'd take a while to switch things over, since I did it all without that module in 2016.

**Comment by KBMD** [comment]
Yes. No point trying to do this retrospectively, but tracking it like you
track points, for future sessions, will be great.


Kevin J. Black

**Comment by KBMD** [comment]
Switching to express.js:  the only thing I'm curious about before you do
that is that I believe node.js was specialty intended to allow numerous
simultaneous logins (or something like that). So before changing, check
whether a change would lose functionality.
Love
Dad


Kevin J. Black

**Comment by jonkb** [comment]
Yeah, express is a package for node, so it wouldn't be leaving node.js.

**Comment by KBMD** [comment]
Oh. Got it, thanks.


Kevin J. Black

**Comment by KBMD** [comment]
Done. Thanks!!
